### ISSUE TITLE
Add query class

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/top-users-in-period#4]
This is different than the top users query in core's `wordpoints_points_get_top_users()` function. That function is just about the current top users, so it performs a user query with usermeta. However, in this case we are summing over a period of time, so we are actually performing a query on the points logs. 

Of course, core includes the `WordPoints_Points_Logs_Query` class. The problem is that to use it we would have to pull out all of the values for a given period and then perform the calculation of the sum of points for each user in PHP.

That might be the best way to go at times, but I think that we should give ourselves the option to actually do the sum in the query itself as well. So of course the code for querying for the top users in a period should be abstracted into a single location, and probably a class is the best solution, since things might get fairly complex.

**Comment by JDGrimes** [comment]
## Performance and MySQL `SUM()`

First, I am interested in just exploring the performance of MySQL's `SUM()` function. I found a question on SO involving a static table with 48M rows that was taking 2 seconds to sum a range of the rows. The suggested solution was to [create a second table with counts precomputed](http://stackoverflow.com/a/27007466/1924128) for a certain chunk size.

In our case, of course, the logs table isn't static, it can still grow. However, that doesn't mean that we couldn't still use a chunk table, because generally speaking the logs are static once they get there, i.e., they don't ever get removed (although there might be edge-cases or issues introduced in the future).

The problem is that we don't necessarily know what chunk size would be appropriate. In that suggestion, the ideal chunk size was pointed out as being about the square-root of the size of the table itself. In this case though, we have a growing table, and we will probably often be selecting smaller chunks than that (day/week/month). (But see http://stackoverflow.com/a/2178484/1924128.)

Actually, I guess you could say that our chunks are defined differently than those were anyway, because in that answer they basically corresponded to row counts, whereas in this case we are basing our periods not on the number of rows but on the timespan. So maybe it isn't so much the number of rows we'd need to take into account, as the ideal timespan.

**Comment by JDGrimes** [comment]
I'm guessing that usually we won't be calculating times much less than a day. I suppose that there might be a desire to calculate the hourly points, or something like that, but for the most part that should be pretty easy anyway. Even on a large site, calculating for the last hour probably wouldn't be much of an issue.

And there is something that we can do as an optimization. Instead of deleting the cached counts when a new log is added, we can simply either automatically update the count, or, when the query is run and the count is needed, we can just get the logs that have occurred since we last cached the results. The only time that would be an issue would be when we'd also have to drop old logs off the back of the cache. However, running a query to sum just those old logs (and subtracting the result) would be less expensive than summing them all. So this would still be a useful optimization, at least on large sites.

**Comment by JDGrimes** [comment]
I suppose that the same optimization could also be used with longer periods of time, like months.

However, the way that we are describing it at this point, it would be a per-query cache, rather than chunking. That is, each query would have a chunk of its own, as it were. It wouldn't be able to be shared across queries. Which would become a problem if we are allowing users to run arbitrary queries, rather than just displaying particular preset queries. (Which is actually exactly what we intend to do in the admin, #3.)

For shorter time periods, that probably isn't a big deal. However, for longer periods, it becomes a problem, especially on larger sites. So we probably do want to implement some kind of chunking to be utilized by longer queries, in addition to the optimized cache for the smaller queries.

**Comment by JDGrimes** [comment]
So we do still have the question of ideal chunk size. This should really be informed by the size of the maximum range that is likely to be queried, rather than the maximum size of the table in this case (which as we said above, we don't know, because it can grow indefinitely).

As far as I can imagine, there is no reason to really run queries for a period greater than a year; if there is a use-case, I'd say it is likely rare. At the very least, I think that it is reasonable for us to build this under that assumption for now, although of course flexibility will surely be an asset in the future.

In that case, I'd say it probably makes sense to chunk at the week level by default. Periods less than a week can be fairly easily calculated, generally speaking. And periods of up to a year would not be that difficult to calculate from per-week chunks, since there would only be 52 involved.

Of course, chunking per week would have its drawbacks, since a week doesn't actually go evenly into either the year or the month, the large periods that are most likely to be queried. But then, nor does any period, really, except for the day. I think that chunking per-day is probably too short though, because we'd still have to add up as many as 365 chunks for each user to get their total for a year.

**Comment by JDGrimes** [comment]
What we can do, of course, is allow the chunk size to be filtered so that sites that need to can optimize for the particular queries that they need to make. Perhaps we could even support multiple chunk sizes, if it makes sense.

**Comment by JDGrimes** [comment]
## Using the points logs query

Assuming that we don't add an expensive join to the query to try and take in users who earned no points in a period (#5), the two main elements that are needed in the query that would prevent us from using the points logs query class from core are:

- The `SUM()` of the points column.
- The `GROUP BY` on the user ID column.

This is assuming of course, that we would make those a part of the query, rather than calculating them in PHP. Which I think given the potential optimizations explored above, we'll likely do.

The benefits of using the points logs query would be that it would already support more complex queries, rather than just all of the logs for a user in a period. (For example, we could calculate just the amount that the user had earned from posts, or even a particular post.) It is already integrated with the date query class from WordPress to run the date queries, too (which are really important here).

Of course, that added complexity would degrade performance when used, but that's not to say that it would be unusable, or that there would be no use-cases for it (I think that there certainly would). It just means that it would have to be used with care, especially on large sites. Which really is true already anyway.

So the thought has come to me that possibly we should modify the points logs query class in core, so that it will support this. Possibly we'd even modify the main database query class itself, to support summing fields and grouping by.

The alternative, though, would be to just sub-class the points logs query class instead, and provide these added features in our child class. Perhaps that is the best way to start out anyway, and once we get things working we can ultimately generalize things and merge them into core if we think that they would be useful. And then we won't be strapping the module to the next core release.

**Comment by JDGrimes** [comment]
Doing this in a subclass is actually simpler than I thought it would be. 

I've decided thought that we're actually going to need several tables and thus several query classes.

First, we're going to need a table for the actual points totals for each user for each block. But because we want to be flexible about block length, we are going to also need a table to keep track of the different blocks. So that second table will have the block type and block period for each block ID.

We'll need a query class for the blocks table, to get a list of blocks that are in the period. Then we'll need a query for the block logs table, the one that actually contains the points totals for each user for each block. And then finally, we'll need a query class for the points logs themselves, to get the data for the blocks and fill in the blanks.

I think really though, that instead of making the points logs query class the primary query class, we should let it be separate from the periodic users query. That would isolate the block handling and stuff from being in the modified points logs query class. It also makes the most sense, because we do need to be able to construct direct queries without block handling to get the data to put into the blocks.

**Comment by JDGrimes** [comment]
That does mean that the main query class won't actually be a wrapper of a table in the database, like the others. It will basically be a meta query class, that just utilizes the other query classes.

In practical terms, this means that we have to come up with our own interface, since extending the `WordPoints_DB_Query` class really wouldn't make sense.

**Comment by JDGrimes** [comment]
One question that I found myself asking was whether we really even need to worry about counting, vs just retrieving all results. The db query classes have a `count()` method, but I wasn't sure that it was useful in this case.

However, I guess there is a potential use-case for wanting to just count the results, and that is when you just want to know the number of users that were awarded points during a given period. I can see that as being something that we may want to do.

On further thought though, I wonder whether this is really the best way to do it. That kind of query would be much more performant than the query to get the points totals, because we'd be counting instead of using `SUM()`. As a result, we could just use a `COUNT( DISTINCT user_id )`, which would probably be able to do a [loose index scan](https://dev.mysql.com/doc/refman/5.7/en/group-by-optimization.html#loose-index-scan).

Of course, the query classes from core don't support distinct counts on particular fields out of the box, but it is trivial to subclass, and we could eventually port it back into core. 

Which is all to say, that we probably don't need to use blocks for counting. But still, that's no reason why we can't offer a `count()` method on on main query class. It just means that it might use a simple query, instead of logic similar to what it would do for actually retrieving the results.

**Comment by JDGrimes** [comment]
## Absolute point of reference for block partitions

One issue here is determining where to partition the blocks at. That is, what absolute reference point to partition them relative to.

Ideally, blocks should be able to fall out in increments that will match up exactly with one or both ends of the query. A fair likelihood of matching up with at least one end of the query would be good. So, regardless of block length, at what point we break the blocks up will come into play there.

For example, even if we were always going to be querying in one-day periods, breaking our blocks up into one-day periods wouldn't necessarily align with the period being queried, because the day might be measured from a different point in time (six o'clock instead of twelve, for instance). This is true regardless of block length.

Assuming that generally we're going to be bringing blocks into play for queries that run from one date to the next, being able to align at least one end of the period with a block sometimes seems like a reasonable expectation, depending on the exact length of the period. Ideally, it would always happen on both ends, but at the very least it should be *possible* for it to happen on one or both ends.

But regardless of block length, if we choose a wrong absolute reference point for where we break the blocks up, it might not be possible for the blocks to ever align with either end of the period. Which again, seems stupid. So ideally we'd like to choose to make the blocks a certain length measured in days, and break them up into those lengths relative to the timezone that the queries will be in.

Usually, we'd expect that the user would want to query relative to the site's timezone. But issues come in when we consider that that can change. And it is possible for it to vary among users as well, if using a plugin or something.

So I'm not sure that we can really design this in a sane manner and also ensure that it is always possible for the blocks to align. There is no one absolute reference point that will work. Despite the mental anguish that not doing this will cause.

However, the changing timezones could be seen as just an edge-case, and we could just optimize for the most likely scenario, which is that once we are installed, the timezone will stay the same.

This does mean though that we can't just base the absolute reference point on the site's current timezone, in case that changes. It would be something that we'd have to set in the database.

**Comment by JDGrimes** [comment]
## Synchronous saving of block data on query?

Another issue we have to consider is how to respond when the block logs have not yet be filled in for some or all of the blocks within a particular period.

Obviously, when that is the case, we need to fill in the block logs for that period. But I'm not sure whether that should be done synchronously or not.

For the purposes of reducing overall query load, filling in the block logs ASAP is the best. Although, I suppose with caching, depending on how we do the caching, it might not really be imperative.

On the other hand is the desire to get the results back to the user ASAP. And for that purpose it is best to just pull out the results across all of the unblocked logs at once. The read query will probably still be significantly faster than the write query (I could be wrong there though, depending on the limits on the number of results the query returns). But in this case, we wouldn't be able to even write the data into blocks if we wanted to, because we'd be adding it together all in one go.

A sort of compromise would be to read the data in blocks, but then cache it in an option while we wait for it to be saved asynchronously. That would mean we wouldn't have to wait for the write query to complete, but it would mean we'd probably be running several read queries, one for each block that needs to be filled in. And that would probably still take longer than just running a single query.

I kind of lean toward this last solution as the most elegant, although perhaps some actual testing would give us some guidance here.

**Comment by JDGrimes** [comment]
## Caching

I think that something that I'm not quite thinking of correctly is the amount of data that we actually end up with after we've run the queries.

Say we run a query to find the top ten users in a period. We end up with 10 user IDs and 10 total values. It is a trivial amount of data to be caching, even for "Top 100" queries.

When it comes to caching though, the issue will be in determining what to cache and what not to. If users on the front end can run any kind of period query, then caching all of that *at the period level* is probably not necessary. Only queries that are likely to be repeated need to be cached.

However, even those one-off period queries can be benefited by caching the totals at the block level. And even there we are talking about a small amount of data to be caching, assuming the number of queries isn't outrageous and the number of users per query is reasonable. Even 100 one-off queries affecting 100 users each, if the queries spanned 5 blocks on average, would yield a block cache size of only 50,000 pairs of integers. Assuming that the integers average around 65,000 or less, each one will take up only two bytes. Which would only be about 98kb. Which is really pretty good, considering this is a fairly liberal scenario.

So the takeaway is, that we probably shouldn't be shy about caching by default.

**Comment by JDGrimes** [comment]
An issue, of course, will be reliably updating/invalidating caches when the logs change. I say updating, because I guess we could just add the values to the existing query when new points are awarded, instead of invalidating the cache completely. At the very least, old blocks could remain cached indefinitely and not go stale, though that could change in the future. 

The problem with just updating the cache for active blocks though, is the presence of limits. Limits on the number of users queried means that not all users are included in the cached results. So a user who received 5 points might not have been in the top 10 users, but now they would be.

That is something that isn't easy to detect, we'd have to run a query on that particular user for the period. Which might not be that expensive, but depending on the number of results in the original query, and how often it is run, invalidating the cache might be a better strategy.

**Comment by JDGrimes** [comment]
This also applies to old blocks if we don't actually include all users in them at all times. Which would mean, if we did, that we wouldn't be able to save them from the query as proposed above, since queries will almost invariably have limits.

However, we would be able to still utilize that data, that is we would be able to save it, but we would need to also get the data for the other users to save as well. Performance-wise, that might not be so beneficial though, possibly we'd be better off just grabbing all users at once.

**Comment by JDGrimes** [comment]
## Blocks and limits

It also highlights a similar issue that we will have with our queries. Because our queries will sometimes be split over the block logs and the regular logs, we wouldn't really know how many users we need to include in the query in order to fulfill the requested limit. That's because each of those queries is independent, and thus applying the limit to both will cause some users to appear in one and not in the other.

If we just based which users to query in the logs on the blocks, we'd still possibly end up with the wrong results, because a user not included in the top users based on the blocks could have earned enough points on the ends queried with the logs query that they would have been in the top users.

Even pulling all the top users from the block logs, plus an increased limit to include that additional number of excess users from the regular logs, would not be sufficient. It would give us a list of top users from just the points logs, but we'd have to go back and query the blocks again in order to get the totals for the users who might have made it into the top just from their regular logs earnings outside the blocks. 

(We actually wouldn't have to check all users, because the amount needed to make it into the top could have increased by the top users from the blocks having their non-block points added. Any users that were only found in the logs that received less that the least amount from those in the blocks, could not possibly make it into the top users. So we could exclude them. Probably we could also do some complex calculation to determine how many users we'd actually have to query based on the limits and how many users could actually receive more than the block users, based on the same principle.)

This means that we'd have to run a third (or forth) query, in addition to the block query and logs query needed for one (or likely both) ends not covered by the blocks perfectly.

This seems like overkill, and makes me wonder how much we are really saving by the use of the blocks. Possibly it would be a good idea to make the periods query smart enough to only have the blocks kick in at a certain point, probably filterable, but based on our tests of when it actually starts to save time/memory/etc. by default.

**Comment by JDGrimes** [comment]
## Reference Point and Multisite

In the discussion about the absolute reference point for the block partitions, I was only thinking of single sites. Or assuming that on multisite there would be one reference point for the network. However, on multisite each site on the network can have its own timezone. Which means we can't just have one reference point for the entire network if want to try to base it on the site timezone. That is, we can't when we are network active, anyway. 

And I doubt that it would really be worth the trouble then. I don't think it would be feasible when it comes to network logs queries especially.

Which is really something that I hadn't considered before, and it leads us to the question of how to handle multisite in general.

I suppose that on multisite the blocks need to be calculated per-site, but then it should also be possible to query network wide, when network active.

Really, I guess every set of query args would require the blocks totals to be stored separately, and this might be no different in terms of implementation. Thought we will need to keep in mind that the site and network a query is for are set automatically, if not supplied by the args.

**Comment by JDGrimes** [comment]
So, given that we store the log dates in GMT, it is probably going to be simplest to just use a UTC reference point by default. For future flexibility though, we could let each block type supply its own reference point. That would allow the blocks to be optimized per-site as necessary.

**Comment by JDGrimes** [comment]
----

In regard to calculating the blocks for all users vs just those top users that have been queried, the benefit of having as few users as possible in the blocks would be increased performance. Although of course, if we have to sacrifice that for accuracy, then we just have to.

**Comment by JDGrimes** [comment]
We could get around the limits issues by running a join on the block logs and the regular logs. But that would potentially be worse for performance. I guess in the end we just need to do some tests.

**Comment by JDGrimes** [comment]
## Query Args

Beyond just the limits and limit related args, we need to consider what other args need to be utilized in the block logs query.

And really, this brings up an interesting issue: that not all of the args supported by the points logs query would be supported by the block logs query.

This is because the block logs only have the points and user columns. And really, the points column means something different than the `points` query args would mean in the context of a regular points logs query. The points query args are meant to reference a transaction's points, whereas in the block logs table it is the total points for the block.

So in reality, only the user-related query args have a direct applicability to the block logs.

The implication being, of course, that if we want to utilize other query args, we would need to not use the blocks. We'd have to just use our points logs query directly.

Or, I suppose an alternative would be to make the block logs also have a column for query signatures. That would allow blocks to be calculated independently for several different queries over the same period, that is, over the same block. That is, the block totals would be calculated for each query independently.

We'd have to do something similar for caching anyway.

And in addition, we wouldn't actually have to add a new column to the block logs table, just to the blocks table itself. So we'd still be able to query based on the block IDs, as we would otherwise, and use the same index, etc.

**Comment by JDGrimes** [comment]
There are some args that should be ignored in the query signature though. The `limit` and `start` args of course, and also the `order` and `order_by` args, since none of these affect the values of points totals, only which users happen to be cached/"blocked".

I guess `fields` is also unnecessary, if we even allow that to be set.

And the `user_*` args don't affect the values of the points totals either, since we are calculating them per-user.

**Comment by JDGrimes** [comment]
That makes me wonder though, whether we ought to assume that they will be calculated per-user at all times. Might not a user want to display the total number of points earned in the last period on a site? But then, that would be a points totals query, and not a top users query. It is a similar concept, and might benefit from the same strategies that we are employing here. But it is really a separate thing, in the sense that it would need to use a different block logs table (without the user column), and thus a different query, etc.

So the general points total queries could be based on the same things that we are building here (maybe even as part of this module?), but we can't really make the top users portion generic in a manner that they can both use the same tables and the same queries.

**Comment by JDGrimes** [comment]
An issue that we mentioned above is that there are some default args of the query, and these also need to be included in the signature. I think the best way to handle that is to construct a points logs query with the args, and then get the list of args from it. However, it turns out that we don't offer a `get_args()` method, just `get_arg()`, which isn't much good. We should remedy that, but really, the only default arg that would be taken into account in the signature anyway is `text__compare` being set to `LIKE`. So for now we can just incorporate that into the signature as needed.

**Comment by JDGrimes** [comment]
That, "as needed" brings to the front another issue, and that is, that we should optimize the query args before we create the signature. They need to be ordered the same each time, and anything superfluous removed. Any sub-args that are arrays should be treated likewise.

**Comment by JDGrimes** [comment]
I was going to say earlier that we might actually need to add a points type column to the block logs table. But the addition of the query signatures takes care of that. I think it is the better solution, because it makes it easier/more performant to query for points across all types, than it otherwise would be with the added column. Which is a potential use-case on some sites. So this is the best solution.

**Comment by JDGrimes** [comment]
## Variable Length Blocks

One thing that I've happened to think as I've been working on things is that on some sites there might be a desire to optimize the block logs by using variable-length blocks.

For example, a site may utilize queries that are based on calendar months, rather than on fixed-length periods (like 30 days). In that case, blocks of exactly 30 days would not be optimal, because they would not only not match the exact length of every month, but they would because of this get desynchronized completely over time. Eventually, the block would start/end in the middle of the month, which would be the least optimal position. And if single-month queries were common, the blocks would actually be totally worthless.

So it seems to me that we might want to consider the possibility of supporting variable-length blocks, even though we might not use them by default. They would be more of an optimization feature that could be applied per-site as needed (and on many small sites would probably be unnecessary).

In reality, no matter what block size is chosen, in relation to the calendar it will not be constant and thus will not stay in sync. Even days can change length relative to a particular timezone, because of daylight savings. 

Of course, we are currently planning to stick with GMT by default, so we don't have to worry about that much (only the occasional leap-second). It is more of an "optimizing for the site's timezone" problem.

**Comment by JDGrimes** [comment]
But if we wanted to go ahead and support that kind of optimization out of the box from version one, I suppose there are several things that we'd have to do.

First, we'd need to always save both the start and end times for each block. This is the way that I had originally designed the database, but I began questioning the necessity of it for fixed-length blocks (which is partly why we are here). 

Secondly, we can't just have the block objects have a `get_length()` method that returns a static value. They'll have to be able to calculate the returned value based on which block in particular is being referenced. So they'll need a block ID or reference time or something to be passed into that method.

And then of course, there will be added complexity relating to the blocks queries, like determining exactly how many blocks we need, which logs to take into account when calculating the values for the blocks, etc. 

Actually, as far as the number of blocks needed, I guess that we don't really need to know that, we just need to know where the closest block break is to the start and end of the period. But after we determine which block that is, we won't know its length until we ask for it.

**Comment by JDGrimes** [comment]
## Caching

Like the blocks, the cache will also be using query signatures. However, because there are some args that are ignored (https://github.com/WordPoints/top-users-in-period/issues/4#issuecomment-289910225) when the block args are calculated, the signatures for the query cache might be different.

The real question would be how we want the cache to operate. If we just want it to cache the results for an exact query, then we can do that. But this will mean that a query with a different limit, or a different order, would be cached separately. This could potential lead to duplicate data within the cache.

The points logs cache operates in a manner that it stores all of the data together, and when a limit is supplied it just truncates the results in the cache, and so on. 

The downside of that approach is that it means that some processing needs to be performed on the cached results before they are returned. But really, it is just a matter of checking whether there are the desired number of results, and then filling in those missing or only returning the number needed.

So I think it makes sense to leave the `limit` and `start` args out of the cache signatures.

However, when it comes to the `order` and `order_by` args, I don't think that these can be handled in a sane manner except for using independent caches.

In the case of the `user_id*` args, I suppose that in some cases we'd be able to pull from the cache based on them, and in other cases we wouldn't (like when using a different comparator than `=`, or using `user_id__not_in`, etc.).

The problem with adding the results of those one-off user queries to the general cache for that query signature though, would be similar to the issue with the order args, and that is that it would create holes within the cache. Which would make it difficult to get the top x users, because the first x cached results wouldn't necessarily match that. I guess really though, that holes could be left in the cache even from just adding the results of queries using the `start` arg. Which is why we end up caching the pages independently in the points logs query if the results aren't already a part of the main cache.

The independent caching also allows the cache to be cleared independently, though this isn't really such a big deal in this case, I think. But it also means that the amount of stuff that has to be pulled into memory when bringing up the cache is smaller, and also, if the cache is stored in the DB, it means smaller read and write operations for the cached results.

So I don't think that we need to worry about getting absolutely optimal caching right off. In fact, there's probably no such thing across every site. For some sites memory is more important, for others, speed/something else. So for now let's just do what is simplest, but make it adaptable, so that we can make it smarter later if needed.

**Comment by JDGrimes** [comment]
## Variable Block Length Support

As far as supporting variable block lengths goes, it can be pretty easily achieved by adding a method that is given a timestamp and returns the start, end, and a length of the block that that timestamp falls within. Really, I guess the length could be easily calculated from the start and end, so maybe it is unnecessary.

Implementing this actually has the effect of simplifying much of the query logic, because the calculations in regard to block lengths, etc., is centralized in a single location (although it might be duplicated across some block types, but this can be fixed by creating a subclass to extend). So we are in a way benefited by supporting it, in addition to the possibility that we'll actually use the feature for optimization in the future.

It also means that the `get_reference_point()` method is no longer necessary, since the reference point is only needed within the block type, not external to it.

As far as I can see, the only issue we would ever run into with this arrangement is if we absolutely had to be able to predict block lengths for some reason. But I don't see why we'd need to be able to do that absent a reference timestamp. So I think this is the way to go.

**Comment by JDGrimes** [comment]
---

As far as not including all users in the block totals, only the ones needed: the problem is that we have to have some way of knowing which users aren't included in which blocks. Because the "top x" may be different for every block, if we only store the top x per block the total across blocks doesn't really tell us anything. At least, that is, it doesn't tell us what we want to know, because it doesn't actually tell us who earned the most points during that period across several blocks.

Because of this, it will likely be simpler just to store the totals for all users for each block in the block logs. Unless we can come up with some ingenious way of knowing which users we don't have totals for from each block, and a good way to get those totals. But as far as I can see, that will just result in excess complexity and excess queries.

**Comment by JDGrimes** [comment]
## Timezones

Even though we are storing all of our times as GMT MySQL `DATETIME` strings, we still have to worry about proper timezone handling, because the start and end dates may be coming from user input based on the sites current timezone. 

Currently I've built the query so that it expects the start and end dates to be supplied as Unix timestamps. This means that they are not timezone relative, they are UTC (which is GMT).

Converting that to the proper format for use in the query is then pretty easy, we can actually just pass the timestamp in directly as a string, just by prefixing it with `@`. Here's the long proof of why that is.

Our queries ultimately end up using `WP_Date_Query`, which doesn't make any explicit statements in its inline docs about timezone related issues. Really, any timezone relating to the query is entirely determined by the dates that are being queried: if the dates are GMT, then your results will be returned as if you had passed query parameters relative to the GMT timezone, because that is what you are querying against. And so for every other timezone. 

The only place where the query logic does PHP-side time calculations is in `build_mysql_datetime()`:

```php
		$now = current_time( 'timestamp' );

/* ... snip ... */

			if ( ! is_array( $datetime ) ) {
				// @todo Timezone issues here possibly
				return gmdate( 'Y-m-d H:i:s', strtotime( $datetime, $now ) );
			}
``` 

That logic kicks in whenever a string is passed, and it does not match any of these formats: `Y`, `Y-m`, `Y-m-d`, `Y-m-d H:i`. Which means it will likely be triggered by us, since we'll probably just pass in a string either `@`-timestamp format or full MySQL `DATETIME` format.

The utilization of `strtotime()` does affect the timezone, because it utilizes the server's default timezone, which WordPress sets to GMT. The `$now` part is based on the site's timezone, but is only utilized for relative strings (e.g., `'+1 day'`), so can be ignored by us here.

This means that whatever we pass in will simply be interpreted as GMT by `strtotime()`. And since we're planning to be using GMT values at that point, that's great.

The use of `gmdate()` means that the use of the GMT timezone is explicitly maintained in the final result (although of course in WordPress `date()` also gives GMT dates by default, since as we said above that is the default timezone).

Thus, when supplying a value broken down into an array to `WP_Date_Query`, timezone is irrelevant, but when supplying a string the result actually utilized in the query will be GMT.

So we do indeed need to be passing GMT values in, and storing GMT values in the database, if we want that convenience of just passing strings to our date queries.

**Comment by JDGrimes** [comment]
So, it makes sense that we want to just accept Unix timestamps for our start and end dates, since these are UTC. The issues come in when we are converting a user-supplied string based on the site's timezone to the correct Unix timestamp.

There is a strong urge to utilize the time handling functions and tools supplied by WordPress, but in fact it is much easier to use a `DateTime` object. Not only is that less complex, but it is also much more accurate when dealing with times supplied by the user, since DST, etc., is not taken into account by WordPress's time functions. (So much so that the way that `current_time()` utilizes offsets is really almost a broken implementation.)

I've created a gist containing a unit test that demonstrates [the difference between `strtotime()` and `DateTime`](https://gist.github.com/JDGrimes/3b34c8ce4c79cbe3dcb2f05486458e3a).

**Comment by JDGrimes** [comment]
This has made me wonder then, whether we ought to just require that the start and end dates be passed in as `DateTime` objects, rather than just using them outside for conversion.

The benefit would be that it would also simplify other calculations and such as the start and end dates get passed around, though mostly that is theoretical calculations that would occur outside of the class itself, when filters were applied or something.

So I guess I can't really pinpoint a good reason for it at this point, other than maybe that it gives us confidence that we are using a properly calculated timestamp, since we'll be able to set the timezone to UTC explicitly before retrieving a timestamp from the object. (It would still be possible for folks to mess up by not supplying the correct timezone when they construct the object though, but that's a separate issue.)

It also means that we wouldn't have to worry about validating (or not validating) the values passed in, since we'd be able to type-hint for the `DateTime` object, whereas anything could be passed in as a "timestamp".

It would also give us the option of passing them around more in the future for other calculations if needed, so maybe it something that we should do. Although note that before messing with the timezone in a filter function, we'd want to clone the object, and not modify the canonical one.

**Comment by JDGrimes** [comment]
> it gives us confidence that we are using a properly calculated timestamp, since we'll be able to set the timezone to UTC explicitly before retrieving a timestamp from the object. (It would still be possible for folks to mess up by not supplying the correct timezone when they construct the object though, but that's a separate issue.)

Actually, changing the timezone value won't change the timestamp at all, since timestamps are always UTC. Whether the timestamp is correct will depend entirely on the object being constructed with the correct timezone, so that the `DateTime` object references the right moment in time.

That doesn't invalidate other benefits though, like being able to grab the formatted value for whatever timezone we want easily.

**Comment by JDGrimes** [comment]
The timezone of a site is saved in the `timezone_string` option, but it isn't always set, so in that case a [work-around using the `offset` option](http://wordpress.stackexchange.com/a/198453/277570) has to be employed. So we'll add a function to make that easier.

**Comment by JDGrimes** [comment]
## `JOIN` 

I'm now wondering whether we should just join the blocks and logs tables, instead of querying them separately. There is no way to limit the results to just the top x users effectively without calculating the points of totals of all users for both the logs and the blocks. Consider:

User | Block Points | Log Points | Total
---- | ------------- | ----------- | -----
1 | 100 | 5 | 105
2 | 7 | 100 | 107
3 | 55 | 57 | 112

User 3 has the most points in this case, but if we are just looking for the top user, for example, they'll be overlooked in both the blocks and the logs, since they don't have the most points in either.

The issue we have is detecting these users who won't appear in the top in either set, only when the totals from both are combined.

Optimizing for this could (possibly, it really needs more thought) be done by finding the top amount from both and then pulling out all users with points that have more than half that. But this can hardly be called an optimization when it may require us to transfer a much larger amount of data from the database, when a join would have performed fairly simple calculations db-side more efficiently.

We have to recognize here that there is really no one-size-fits-all "best" strategy here. Different servers and different queries would have different ideal solutions. But I'm thinking that due to this problem, a join will likely be either the best or a reasonably efficient solution for most cases, rather than pulling the data out and performing an optimized join PHP-side.

This doesn't mean that the blocks aren't useful at all though, since the decrease the number of rows that have to be summed. For longer periods, I think the join with the blocks would still be more efficient than just a straight query involving only the points logs table. But some testing might elucidate this further.

**Comment by JDGrimes** [comment]
This [SO answer](http://stackoverflow.com/a/2387766/1924128) explains what kind of query we need to run here:

```sql
SELECT id, SUM(amount) FROM
(
    SELECT id, SUM(amount) AS `amount` FROM t1 GROUP BY id
  UNION ALL
    SELECT id, SUM(amount) AS `amount` FROM t2 GROUP BY id
) `x`
GROUP BY `id`
```

A basic version of this using our tables gives this `EXPLAIN`:

![vvv dev_localhost_wordpress_unit_tests_phpmyadmin_4 2 11_-_2017-04-04_08 50 22](https://cloud.githubusercontent.com/assets/4005415/24658189/fd57a51a-1916-11e7-925f-6e4a19327c80.png)

Which as far as our tables go is nearly ideal. The real issue of course will be the performance of the primary query, since it is using a temporary table and filesort, but I'm not sure that there is any way to improve it.

**Comment by JDGrimes** [comment]
## Filling in the block logs

I've been looking into how to best fill in the block logs. I had always kind of envisioned us pulling out the logs data and then looping over it to insert it one by one. However, now that we are going to try using a union query, we know longer need to pull up the data for each block as a part of calculating the totals; due to the limit everything has to be included in a single query.

So I've been thinking that instead of pulling out all of that data and then stuffing it right back in, it would be better to just use an `INSERT ... SELECT` statement. I have tried to investigate the potential pros and cons of this vs pulling the data out, and I haven't really found any cons. The benefit, of course, is that it is all done db-side, instead of sending all of that data around.

Specifically, one of my worries was [table locking](https://dev.mysql.com/doc/refman/5.7/en/table-locking.html), although this shouldn't be much of a problem (especially for `InnoDB` tables), since we are neither reading nor writing to the table very often.

**Comment by JDGrimes** [comment]
We've debated above whether it was better to fill the blocks in inline while we are performing the query, or to wait until later, and the need to do a single query also has a bearing on that.

It is possible for us to construct the query in such a way that it would still include all of the points logs needed, even if there were just some holes between the blocks (some blocks missing in the middle), but this increases query complexity, and potentially would have a considerable impact on performance.

The question that we've asked before and still need to answer is, will that performance impact sometimes be worse than if we'd just created the blocks inline? And if so, at what point?

I think really though, that we can pretty emphatically answer "no" to the first question under usual circumstances. That is, the performance perceived from the perspective of the person running the query will almost always be better when we don't fill the blocks inline. The overall performance of the site will be less though, because we still have to come along and fill those blocks in at some point. And while we're getting around to that, we might find ourselves running several more queries that would have been improved by the existence of those blocks. So it seems superior to fill the blocks inline in terms of overall performance, though this may affect the initial performance of the primary query in question to a degree.

The argument for simplicity would also point us in the same direction.

The main thing worth noting though is that when we are displaying the results in a widget or something, the user has not consciously initiated the query. It is one thing when an admin is performing a query via the admin screen and has to wait a few extra seconds (we could even explain why it might take a little longer); it is quite another when a user has no idea why they are waiting.

We could, of course, circumvent the issue altogether by just filling in all of the blocks beforehand. However, that isn't ideal, especially since we may be dealing with multiple block types. It would lead to more data ending up in the block logs, and would also mean that our queries wouldn't work at all until all of the blocks were calculated. Which is an issue, so no.

As far as the widget goes, I suppose that if the Customiser is being used, the query will have to run just to show a preview anyway. So the blocks will generally have been filled in before the table is actually displayed to site visitors.

**Comment by JDGrimes** [comment]
Whether we fill out the blocks inline or not, we face a kind of conundrum when it comes to insuring that the block is not used in any queries until it is full. If we insert the block into the blocks table first, then a second query running simultaneously will think that the block is already filled in, and will attempt to use it. Except that it won't be filled in completely until the insert query is finished. So to avoid race conditions, we may need to lock the table or something anyway.

Just filling in the block logs first doesn't fix the issue, because then a second query would think that the block did not exist (since it wouldn't be in the blocks table) and would itself being trying to fill it in.

One option would be to insert the block first but mark it as a "draft" in some way, so that other queries would know that it was being filled in and not to touch it. However, this also means that those other queries then have to stand by and wait for the query to finish, or else bail out. Which I guess really they'd have to do anyway, regardless of how we attempt to prevent race conditions.

But rather than bailing or continuously re-checking the blocks status while we wait for it to be filled, it would be simpler to just leave this in MySQL's hands. Which we might be able to do by locking the table in some way. Then we'd be able to just insert the block directly, without the need to mark it as a draft, since a query that tried to use it wouldn't be able to run until the data had finished inserting into the block logs table. Actively waiting in line on our part would no longer be necessary.

I wonder though whether this could partly be dealt with at a higher level, that is, at the cache level. If two queries are using the same block simultaneously, there's some chance that they are actually the same query being run by two separate sessions. Instead of running that query twice, it should just run once and the cache be filled and then utilized by the other query. I think pursuing some kind of thing like that is probably worth it in this case because we have such a potentially resource intensive process here. Just think about what would happen if a new widget was added on the Widgets admin screen for a very active site. Tens or hundreds of copies of the query could end up getting issued before the cache was filled out. And the problem would be a vicious cycle, as those subsequent queries ate up resources causing the initial one to run even longer. So some way of "query once, read cache twice", would be almost a necessity here.

And depending on the situation it might be better (and certainly simpler) to just display a message that the page would need be reloaded after a few moments to see the results, instead of waiting for the cache to be filled while we repeatedly re-check it.

**Comment by JDGrimes** [comment]
Or, instead of requiring the page to be reloaded, we could fill the results in via Ajax. But I don't think that we have to pursue something like that in the initial version.

**Comment by JDGrimes** [comment]
## Conditions on `total`

I had assumed that it would be possible to have conditions on the total, like just getting all users that earned more than 1000 points within a period. However, there is a catch. The aggregated values cannot have conditions in the `WHERE` clause (makes sense really), but [by adding a `HAVING` clause after `GROUP BY`](http://stackoverflow.com/a/16517352/1924128). 

This will take extra customization of the query classes to handle, but I suppose that it is probably worth it.
